write-fonts/src/tables/cmap.rs

```diff
@@ -237,6 +236,187 @@ impl Cmap {
         ))
     }
 }
+// a helper for iterating over ranges for cmap 4
+struct Format4Ranges<'a> {
+    mappings: &'a [(char, GlyphId)],
+    seg_start: usize,
+    currently_contiguous: bool,
+}
+
+#[derive(Clone, Copy, Debug)]
+struct Format4Segment {
+    // indices are into the source mappings
+    start_ix: usize,
+    end_ix: usize,
+    id_delta: Option<i32>,
+}
+
+impl Format4Segment {
+    fn len(&self) -> usize {
+        self.end_ix - self.start_ix + 1
+    }
+
+    // cost in bytes of this segment
+    fn cost(&self) -> usize {
+        const BASE_COST: usize = 8; // 4 u16s for a new segment
+        let glyph_id_cost = self
+            .id_delta
+            .is_none()
+            .then(|| self.len() * u16::RAW_BYTE_LEN)
+            .unwrap_or(0);
+        BASE_COST + glyph_id_cost
+    }
+
+    fn combine(&self, other: Format4Segment) -> Format4Segment {
+        assert_eq!(other.start_ix, self.end_ix + 1);
```

Comment:
Not loving the assert, I suggest modelling the failure explicitly, e.g. try_combine or similar.

Reply:
https://github.com/googlefonts/fontations/pull/1230 explores this idea

Reply:
this is just documenting an invariant, we only combine contiguous segments.  Can add a comment...